test/sqllogictest/dates-times.slt

```diff
@@ -1368,3 +1368,106 @@ query T
 SELECT now() + null;
 ----
 NULL
+
+# Test timestamp precision.
```

Comment:
needs tests for values outside [0,6]:

```
postgres=# SELECT '1970-01-01T00:00:00.666666'::date::timestamp(12);
WARNING:  TIMESTAMP(12) precision reduced to maximum allowed, 6
LINE 1: SELECT '1970-01-01T00:00:00.666666'::date::timestamp(12);
                                                   ^
      timestamp      
---------------------
 1970-01-01 00:00:00
(1 row)
```

Reply:
So, as per the changes of this PR, if the value is more than 6 it will result in an error. I added more tests

Reply:
I don't think this should error if the precision is greater than 6; it should reduce it to 6 and (ideally) throw a warning. Adding the warning might be challenging to accomplish, so should at the very least be left as a `TODO` + GitHub issue to follow up.

Reply:
Oh, disregard this comment as per https://github.com/MaterializeInc/materialize/pull/21570#discussion_r1318053906